### PR TITLE
Add degree aware FFT

### PR DESCRIPTION
**Pull request by andrewmilson**
## Description

closes: #83

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (master)
- [ ] Linked to GitHub issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [ ] Wrote unit tests
- [ ] Updated relevant documentation in the code
- [ ] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the GitHub PR explorer

### CODE REVIEW

File: poly/src/domain/radix2/fft.rs [hunk 1]
```diff
@@ -20,6 +20,46 @@ enum FFTOrder {
...
+        let num_coeffs = coeffs.len();
+        let log_d = ark_std::log2(num_coeffs);
+        // When the polynomial is of size k*|coset|, for k < 2^i,
+        // the first i iterations of Cooley Tukey are easily predictable.
+        // This is because they will be combining g(w^2) + wh(w^2), but g or h will always refer
+        // to a coefficient that is 0.
+        // Therefore those first i rounds have the effect of copying the evaluations into more locations,
+        // so we handle this in initialization, and reduce the number of loops that are performing arithmetic.
+        // The number of times we copy each initial non-zero element is as below:
+        let duplicity_of_initials = 1 << (log_n - log_d);
```
Comment (Pratyush) [inline poly/src/domain/radix2/fft.rs:40]: Let's make this is a `checked_sub`, or maybe handle the case where `d >= n` separately.

File: poly/src/domain/radix2/fft.rs [hunk 2]
```diff
@@ -20,6 +20,46 @@ enum FFTOrder {
...
+            if i < ri {
+                coeffs.swap(i as usize, ri as usize);
+            }
+        }
+
+        // duplicate initial values
+        if duplicity_of_initials > 1 {
+            for i in (0..n).step_by(duplicity_of_initials) {
+                let v = coeffs[i];
+                coeffs[i + 1..i + duplicity_of_initials].fill(v);
```
Comment (Pratyush) [inline poly/src/domain/radix2/fft.rs:55]: I think this is equivalent? 

It's also parallelizable, so might be a tad bit faster. There are also potentially other parallelizations that could be better (e.g. just chunking `coeffs` up into `num_cpu` chunks, and allocating each chunk to a thread to process serially.)

```suggestion
            ark_std::cfg_chunks_mut!(coeffs, duplicity_of_initials).for_each(|c| {
            	let v = c[i];
                c[i + 1][.. duplicity_of_initials].fill(v);
            });
```

**Comment by Pratyush** [comment]
Thanks, this PR is awesome! If you'd like to add benchmark numbers that would be great. 

For the `unused` errors, you can add `#[allow(unused)]` on the relevant items.

**Comment by andrewmilson** [comment]
@Pratyush  sad to say the benchmarks are rather depressing. ~7% performance regression on the FFT tests. I added a couple of benchmarks for polynomial multiplication. Multiplication of two polynomials with equal degree also has ~7% regression in performance. The only situation this implementation performs better is when multiplying polynomials of substantially different degree i.e. `|poly_a| = |poly_b| * 32` - I actually tested this specific case and it only had a 10% gain over the O(n log n) implementation. 

I think the performance degradation is due to `fft_in_place` using oi_helper (now) instead of io_helper (before). The master branch IFFT benchmarks are 10% slower than the master branch FFT benchmarks and I'm pretty sure it's attributed to IFFTs using `oi_helper` and FFTs using `io_helper`

**Comment by Pratyush** [comment]
Wow that's quite surprising. Let me take a look and see if we can optimize the out_of_order code.

**Comment by Pratyush** [comment]
@ValarDragon any ideas? I've been looking, but can't find a fix.

**Comment by Pratyush** [comment]
@andrewmilson I've changed it to only invoke the degree_aware_fft when `domain.size() >= p.degree() * 4`. I'm seeing perf improvements from doing that. Let me know if this looks good to you!

**Review by Pratyush — APPROVED** [review]

File: poly/src/domain/radix2/mod.rs [hunk 3]
```diff
@@ -142,8 +142,17 @@ impl<F: FftField> EvaluationDomain<F> for Radix2EvaluationDomain<F> {
...
+            coeffs.len()
+        } else {
+            coeffs.len().next_power_of_two()
+        };
+        if self.size() >= (num_coeffs << 2) {
+            self.degree_aware_fft_in_place(coeffs);
+        } else {
+            coeffs.resize(self.size(), T::zero());
+            self.in_order_fft_in_place(coeffs);
+        }
```
Comment (andrewmilson) [inline poly/src/domain/radix2/mod.rs:155]: Could this just be

```rust
if coeffs.len() * 4 <= self.size() {
    self.degree_aware_fft_in_place(coeffs);
} else {
    coeffs.resize(self.size(), T::zero());
    self.in_order_fft_in_place(coeffs);
}
```

?

Comment (andrewmilson) [reply poly/src/domain/radix2/mod.rs]: Pushed these changes. Marking resolved

**Comment by andrewmilson** [comment]
Awesome thanks. LGTM

**Pull request merged**